functions/api/report.js

```diff
@@ -0,0 +1,235 @@
+    }
+
+    // Check if branch is valid
+    console.info(`Generating report for branch ${branchId} for period ${fromDate} to ${toDate}`)
+    const branchRef = db.collection(constants.BRANCHES).doc(branchId);
+    const branch = await branchRef.get()
+    if (!branch.exists) {
+        const error = new Error(`Requested branch is not present in Firestore.`)
+        error.statusCode = 404
+        next(error)
+        return;
+    }
+
+    // Add branch info to the report
+    const branchReport = addBranchData(branch, fromDate, toDate)
+
+    // Get Previous and Next Dates for calculation
+    fromDate = utils.prevDate(fromDate)
+    toDate = utils.nextDate(toDate)
+
+    // Get current inventory for the given branch
+    const branchInventory = await getInventory(branchRef, branchId, false)
+    const inventoryProducts = branchInventory[constants.INVENTORY]
+
+    // Calculate report for each product in inventory
+    var totalProducts = 0
+    const promises = []
+    for (const product of inventoryProducts) {
+        const productPromise = getReportForProduct(branchRef, product, fromDate, toDate)
+        promises.push(productPromise)
+    }
+
+    // Wait for completing all promises
+    Promise.all(promises)
+        .then(reports => {
+            reports.forEach(productReport => {
+                if (hasTransactionrecordBeforeToDate(productReport)) {
+                    branchReport[constants.REPORT].push(productReport)
+                    totalProducts = totalProducts + 1
+                }
+            })
+            branchReport[constants.TOTAL_PRODUCTS] = totalProducts
+            response.status(200).json(branchReport)
+        })
+})
+
+/**
+ * Method to check if product had a transaction before toDate
+ * @param {*} productReport 
+ */
+function hasTransactionrecordBeforeToDate(productReport) {
+    if(productReport[constants.INITIAL_QUANTITY] == 0 
+        && productReport[constants.CLOSING_QUANTITY] == 0) {
+            return false
+        }
+    return true
+}
+
+/**
+ * Utility method to add branch data
+ * @param {*} branch 
+ * @param {*} fromDate 
+ * @param {*} toDate 
+ */
+function addBranchData(branch, fromDate, toDate) {
+    const branchReport = {}
+    branchReport[constants.BRANCH] = branch.id
+    branchReport[constants.NAME] = branch.data()[constants.NAME]
+    branchReport[constants.FROM_DATE] = fromDate
+    branchReport[constants.TO_DATE] = toDate
+    branchReport[constants.REPORT] = []
+    return branchReport
+}
+
+/**
+ * Method to generate report for a given product for a particular branch
+ * @param {*} branchDocument 
+ * @param {*} product 
+ * @param {*} fromDate 
+ * @param {*} toDate 
+ */
+async function getReportForProduct(branchDocument, product, fromDate, toDate) {
+
+    let productTransactionRef = await branchDocument
+        .collection(constants.TRANSACTIONS)
+        .where(constants.PRODUCT, '==', product[constants.PRODUCT])
+        .where(constants.DATE, '>', fromDate)
```

Comment:
not including fromdate and todate ?

Reply:
I am. in this method fromDate is fromDate -1 and toDate is toDate + 1

Reply:
@shrikantdange1208 isnt fromDate meant to be inclusive. so in that case 2020-07-18 => >= 2020-07-18 00:00:00

Reply:
This fromDate is fromDate - 1 and toDate is toDate + 1